engine/src/w32printer.cpp

```diff
@@ -1986,7 +1986,7 @@ bool MCWindowsPrinter::FetchDialogData(HGLOBAL& r_devmode_handle, HGLOBAL& r_dev
 			t_devnames -> wDefault = 0;
 
 			// SN-2014-08-07: [[ Bug 13084 ]] The pointer arithmetic wasn't right
-			MCStringGetChars(*t_string, MCRangeMake(0, MCStringGetLength(*t_string)), (unichar_t*)(t_devnames + 1));
+			MCStringGetChars(*t_string, MCRangeMake(0, MCStringGetLength(*t_string)), (unichar_t*)(t_devnames + sizeof(unichar_t) * 1));
```

Comment:
this doesn't look right - the old code was adding 1 to t_devnames which is a pointer to DEVNAME resulting (in bytes) in t_devname + 1 * sizeof(DEVNAME). From the docs https://docs.microsoft.com/en-us/windows/win32/api/commdlg/ns-commdlg-devnames this looks like the correct thing to do - paste the string after the struct.
I think the error is in the values added to wDriverOffset and wOutputOffset which don't take the size of unichar_t into account
```
t_devnames -> wDriverOffset = t_devnames_size - 1 * sizeof(unichar_t);
...
t_devnames -> wOutputOffset = GetDeviceOutputType() == PRINTER_OUTPUT_FILE ? t_devnames_size - 6 * sizeof(unichar_t) : t_devnames_size - 1 * sizeof(unichar_t);
```

Reply:
@livecodeian Hmm I think I had tried the suggested change and it did not work - let me try this again.

Reply:
Yeap, the suggested change does not seem to fix the problem.

Recipe:
```
on mouseUp
   answer printer // choose a printer other than the default
   answer page setup
   answer the printerName // --> the default printer is returned instead of the one chosen before
end mouseUp
```

Reply:
There are similar errors in StoreDialogData handling of DEVNAMES which may be why it's still broken

Reply:
@livecodeian where exactly in StoreDialogData handling of DEVNAMES? I cannot spot anything wrong

Reply:
Actually it's fine as is - I misread the API docs. The offsets in the struct are given in chars rather than bytes so what needs to change is the calculations involving `sizeof(DEVNAMES)` as that returns a number of bytes so:
Line 1977 becomes
`t_devnames_size = sizeof(DEVNAMES) / sizeof(unichar_t) + MCStringGetLength(*t_string);`
Line 1985 becomes
`t_devnames -> wDeviceOffset = sizeof(DEVNAMES) / sizeof(unichar_t);`

Reply:
Thanks @livecodeian, that makes sense. This change returns correct results with most of the printers. However, there are still printers where if I chose them in `answer printer`, and then call `answer page setup`, they are not saved properly, and calling `answer the printername` returns the default printer instead of the chosen one in these cases. Chosing the same printers works fine in LC 6.x, so it seems that still something somewhere is broken :(

Reply:
Ah! I notice the string construction on line 1971 uses a c-string literal with escaped nulls (`'\0'`) as a format string. That won't work as the formatting will stop at the first null encountered. Instead try constructing a mutable string with MCStringAppend:
```
MCStringCreateMutable(&t_string);
MCStringAppend(*t_string, m_name);
MCStringAppendChar(*t_string, '\0');
MCStringAppend(*t_string, MCSTR("FILE:"));
MCStringAppendChar(*t_string, '\0');

Reply:
This sounds promising @livecodeian :)

I'll try this tomorrow - thanks!

Reply:
Works like a charm, good catch @livecodeian :)